### PR TITLE
node: devicemgr: Allow takeover for device plugin

### LINKED ISSUE #127803: Support takeover for devicemanager/device-plugin
### What would you like to be added?

Add support for seamless takeover by a new device-plugin for a resource without deregistering the resource.

In other words, a new device-plugin can be started and registering the resource before the old device-plugin is stopped. In this case, devicemanager should not change the values of allocatable/total count for this resource.

### Why is this needed?

Currently, devicemanager only works with one device-plugin instance for every resource. In other words, if two device-plugin are running, devicemanager may have unexpected behaviors. 

For example, when two device-plugin are running, they are both registered for the same resource. However, if one of them stops, the resource is marked as deregistered and resource counts reset to 0.

We need this support to have seamless deployment of device-plugin. After this is supported, new device-plugin can be started while the old device-plugin is also running, and devicemanager should notice the new device-plugin, as it registers more recently, to be the new takeover and gracefully deregister/disconnect the old device-plugin. With this, there will be no downtime for resources in device-plugin updates.

### PR DESCRIPTION
**Pull request by linxiulei**
#### What type of PR is this?

/kind feature

#### Which issue(s) this PR fixes:

Fixes #127803

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

### CODE REVIEW

File: pkg/kubelet/cm/devicemanager/manager.go [hunk 1]
```diff
@@ -237,6 +237,12 @@ func (m *ManagerImpl) PluginConnected(resourceName string, p plugin.DevicePlugin
 
 	m.mutex.Lock()
 	defer m.mutex.Unlock()
+	newEndpoint := endpointInfo{e, options}
+	if ep, exists := m.endpoints[resourceName]; exists {
+		klog.V(2).InfoS("Endpoint is tookover", "resourceName", resourceName, "prevEndpoint", ep, "newEndpoint", newEndpoint)
+		ep.e.setStopTime(time.Now())
+	}
+	m.endpoints[resourceName] = newEndpoint
 	m.endpoints[resourceName] = endpointInfo{e, options}
```
Comment (bart0sh) [inline pkg/kubelet/cm/devicemanager/manager.go:246]: What's the reason to assigning the same value twice here?
```
	m.endpoints[resourceName] = newEndpoint
	m.endpoints[resourceName] = endpointInfo{e, options}
```

**Comment by bart0sh** [comment]
@linxiulei please cover this use case with either unit or e2e test(s), thanks!

File: pkg/kubelet/cm/devicemanager/manager.go (hunk 1 above)
Comment (linxiulei) [reply pkg/kubelet/cm/devicemanager/manager.go]: Fixed

**Comment by linxiulei** [comment]
> @linxiulei please cover this use case with either unit or e2e test(s), thanks!

Added.

File: pkg/kubelet/cm/devicemanager/plugin/v1beta1/handler.go [hunk 2]
```diff
@@ -106,8 +106,14 @@ func (s *server) deregisterClient(name string) {
...
 
-	c = s.getClient(name)
-	if c == nil {
+	curClient := s.getClient(name)
+	if curClient == nil {
+		return
+	}
+
+	// A new device plugin client has taken over and became the current.
+	if curClient.ID() != c.ID() {
```
Comment (lengrongfu) [inline pkg/kubelet/cm/devicemanager/plugin/v1beta1/handler.go:115]: For same device-plugin this ID is same, eg. k8s-device-plugin socket value is `/var/lib/kubelet/device-plugins/nvidia-gpu.sock`. how to deal with this issue?

Comment (bart0sh) [reply pkg/kubelet/cm/devicemanager/plugin/v1beta1/handler.go:115]: As far as I understood this PR covers another scenario: new plugin should use different socket.

**Comment by bart0sh** [comment]
@pohly This PR seem to be slightly related to https://github.com/kubernetes/kubernetes/issues/128964

File: pkg/kubelet/cm/devicemanager/plugin/v1beta1/handler.go (hunk 2 above)
Comment (nojnhuh) [reply pkg/kubelet/cm/devicemanager/plugin/v1beta1/handler.go:115]: Could we clarify the semantics of `ID` in a comment on the method definition in the interface?

File: pkg/kubelet/cm/devicemanager/manager_test.go [hunk 3]
```diff
@@ -269,6 +269,61 @@ func TestDevicePluginReRegistrationProbeMode(t *testing.T) {
 	cleanup(t, m, p1)
 }
 
+// Tests that the device plugin manager and client correctly handle the take over scenario
+// that a new device plugin is started while another device is running and managing the
+// same devices.
+// Whle testing above scenario, the resource allocatable and capacity values should stay
+// non-zero and equal.
+func TestDevicePluginTakeover(t *testing.T) {
+	// TODO: Remove skip once https://github.com/kubernetes/kubernetes/pull/115269 merges.
```
[the remark this answers is not in the document]
Comment (linxiulei) [reply pkg/kubelet/cm/devicemanager/manager_test.go:278]: I think https://github.com/kubernetes/kubernetes/pull/115269 was closed by no progress for a while but the intent to fix may still exist. I'll keep as it is until there is clearer plan for that issue.

File: pkg/kubelet/cm/devicemanager/plugin/v1beta1/handler.go (hunk 2 above)
Comment (linxiulei) [reply pkg/kubelet/cm/devicemanager/plugin/v1beta1/handler.go:116]: Added comments for `ID()` and here. Thanks!

File: pkg/kubelet/cm/devicemanager/manager_test.go [hunk 4]
```diff
@@ -269,6 +269,61 @@ func TestDevicePluginReRegistrationProbeMode(t *testing.T) {
...
+	case <-time.After(5 * time.Second):
+	}
+
+	capacity, allocatable, _ = m.GetCapacity()
+	resourceCapacity = capacity[v1.ResourceName(testResourceName)]
+	resourceAllocatable = allocatable[v1.ResourceName(testResourceName)]
+	require.Equal(t, resourceCapacity.Value(), resourceAllocatable.Value(), "capacity should equal to allocatable")
+	require.Equal(t, int64(2), resourceAllocatable.Value(), "Devices shouldn't change.")
+	cleanup(t, m, p2)
+}
```
Comment (bart0sh) [inline pkg/kubelet/cm/devicemanager/manager_test.go:325]: What would happen if Kubelet restarts in the middle of this process, i.e. when both plugins are listening on their sockets for Kubelet connects?

Comment (linxiulei) [reply pkg/kubelet/cm/devicemanager/manager_test.go:325]: Good question, and this is very tricky.

Assuming Kubelet discovers plugin one by one (probably not, but just to simplify the discussion):

case 1:
1. old plugin is connected first
2. new plugin then takes over
3. old plugin is tear down and all is good

case 2

1. new plugin is connected first
2. old plugin then takes over
3. old plugin is tear down, and Kubelet would release the resource
4. Kubelet effectively ignore the new plugin and the resource capacity is zero

Although this is a problem, I think it's pretty rare and solvable by plugin's own implementation: 

The plugin that wants to implement high availability should be aware of its socket path's prefix so they can discover if there is old plugin running before them. The new plugin should delete the old plugin's socket file after creating new socket path to prevent Kubelet from connecting the old plugin as in the suggested case. The deletion of the socket file should not affect the use of established connection to old plugin, so no impacts there. 

WDYT?

Comment (bart0sh) [reply pkg/kubelet/cm/devicemanager/manager_test.go:325]: @linxiulei I'd suggest to also look at https://github.com/kubernetes/kubernetes/issues/128964. It would be great to come up with a solution that works for all 3 Kubelet plugins (CSI, DRA, Device).

Comment (pohly) [reply pkg/kubelet/cm/devicemanager/manager_test.go:325]: @linxiulei: in principle I like the idea of one pod cleaning up after an older one. That also solves the problem of leaving garbage behind when pods die unexpectedly and the replacement pod uses a new, unique socket path. But there is one catch:

> The plugin that wants to implement high availability should be aware of its socket path's prefix so they can discover if there is old plugin running before them.

How does one pod instance determine that it is newer than some other instance?

Suppose a node goes down unexpectedly while the daemonset is updating and there are two pods scheduled to the node. Both pods are not deleted yet. When kubelet comes back online, it starts both pods. It happens that the *new* one runs first, then the *old* one. If each pod deletes all other sockets except its own, we end up with kubelet talking to the old pod and once that finally goes away, there's no active plugin anymore.

Can that happen? If not, why?

It depends on how exactly the daemonset controller does the upgrade. I haven't looked into that yet.

My own idea for https://github.com/kubernetes/kubernetes/issues/128964 was to inject the daemon set spec generation into the pod on the node. That works, there is a `pod-template-generation=<generation number>` label which is supported by the downward API and can be considered an official API - albeit a deprecated one:
https://github.com/kubernetes/kubernetes/blob/e69a5ed9b3764347c485cd4854149f3174d4bd95/staging/src/k8s.io/api/extensions/v1beta1/types.go#L553-L562

With DRA, each plugin instance could register with that generation in a special entry in the "supported versions". The semantic of that array was intentionally defined in 1.32 to allow adding such an entry. Then the kubelet can determine which plugin instance is newer by comparing those numbers. This can be combined with deleting older sockets, if their path includes that generation.

What I don't like about this approach, besides the dependency on a deprecated label, is that it fails when the daemonset gets removed and created again. In that case the newer pods will start with a generation of 1 while the old ones might be the same or higher (if updated at least once). I haven't looked at how the hash is calculated, but I wouldn't be surprised to see that this loophole was why the generation label got replaced.

Another alternative to using unique socket paths is to always unlink the socket (if there is one) and create a new one each time a pod starts. If the kubelet detects that it lost a connection after the old pod is gone, it needs to re-register the plugin under that same socket to get the newer registration infos. I think it currently doesn't do that, but @bart0sh's work on detecting stale connections might cover this aspect. But it doesn't work reliably either if the two pods might get started in the wrong order.

Comment (linxiulei) [reply pkg/kubelet/cm/devicemanager/manager_test.go:325]: > Can that happen? If not, why?

Yes, it can happen. However, it seems a bigger issue out of plugin pod's control . One possible solution in kubelet could be handling DaemonSet Pods specially:

When there are multiple pods controlled by the same DaemonSet and none of them are running, Kubelet should start them sequentially according to their pod creation timestamp. 

This sounds a workaround to me and might have other problems though.

`
Another alternative to using unique socket paths is to always unlink the socket (if there is one) and create a new one each time a pod starts. If the kubelet detects that it lost a connection after the old pod is gone, it needs to re-register the plugin under that same socket to get the newer registration infos. I think it currently doesn't do that, but @bart0sh's work on detecting stale connections might cover this aspect. But it doesn't work reliably either if the two pods might get started in the wrong order.
`

Does it make HA/taking-over possible? It sounds the old pod needs to go first then kubelet would register the new pod, which leaves a short period of no pod registered.

Comment (pohly) [reply pkg/kubelet/cm/devicemanager/manager_test.go:325]: > When there are multiple pods controlled by the same DaemonSet and none of them are running, Kubelet should start them sequentially according to their pod creation timestamp. This sounds a workaround to me and might have other problems though.

I agree, let's not do that. It sounds wrong for kubelet to learn about higher-level app concepts.

Perhaps the pod creation time stamp can be used to determine the "newer" pod. Unfortunately, I don't see how the plugin can get to that information because the [downward API](https://kubernetes.io/docs/concepts/workloads/pods/downward-api/) does not expose it.

> > Another alternative to using unique socket paths is to always unlink the socket
> Does it make HA/taking-over possible?

If the kubelet has the old socket open at the time that it gets unlinked and thus should still be able to communicate with the old pod until it shuts down. Then once the registration of the new pod is complete, the kubelet can switch to the new connection.

There might be an impedance mismatch between gRPC "connections" and Unix domain sockets + pod lifecycle: if I remember correctly, a gRPC connection is about "talk to the server at this address or socket path". If the server restarts, it's still the same connection for gRPC, it was just temporarily down. If the path is the same across pod restarts, we want to track gRPC server restarts (= pod restarts) because on a restart we need to re-register the plugin.

@bart0sh: have you seen https://github.com/kubernetes-csi/csi-lib-utils/blob/master/connection/connection.go and in particular the `OnConnectionLoss` option? I added that in https://github.com/kubernetes-csi/csi-lib-utils/commit/8147e17c70332f5c216c00907e02c2fa24a3a016.

Comment (pohly) [reply pkg/kubelet/cm/devicemanager/manager_test.go:325]: One more thing while we are at it: why does plugin registration use two sockets, one for the generic registration service and one for the domain-specific service(s) (device plugin/CSI/DRA)?

Wouldn't it be simpler to have just one socket where a plugin provides all gRPC interfaces that it supports?

Let me toss out that thought, just in case that it becomes relevant. I'm not sure whether anything can be done without breaking compatibility, but who knows...

Comment (ffromani) [reply pkg/kubelet/cm/devicemanager/manager_test.go:325]: > One more thing while we are at it: why does plugin registration use two sockets, one for the generic registration service and one for the domain-specific service(s) (device plugin/CSI/DRA)?
> 
> Wouldn't it be simpler to have just one socket where a plugin provides all gRPC interfaces that it supports?
> 
> Let me toss out that thought, just in case that it becomes relevant. I'm not sure whether anything can be done without breaking compatibility, but who knows...

FWIW I'm not a fan of the dual-socket model either. The work @SergeyKanzhelev started about device plugins reliability touches also this aspect. Unfortunately I'm skeptical we can change the model without breaking compability, but I'd be interested in alternatives, and willing to help as much as I can (likewise for the device plugin reliability effort)

Comment (bart0sh) [reply pkg/kubelet/cm/devicemanager/manager_test.go:325]: @pohly
> have you seen https://github.com/kubernetes-csi/csi-lib-utils/blob/master/connection/connection.go and in particular the OnConnectionLoss option?

No, I haven't. Thanks for pointing out. So far I'm using [WaitForStateChange](https://pkg.go.dev/google.golang.org/grpc#ClientConn.WaitForStateChange) API.

Comment (linxiulei) [reply pkg/kubelet/cm/devicemanager/manager_test.go:325]: > If the kubelet has the old socket open at the time that it gets unlinked, it should still be able to communicate with the old pod until it shuts down. Then once the registration of the new pod is complete, the kubelet can switch to the new connection.

If I understood it correctly, we still need a mechanism here to switch to new connection while old pod is around so the process has no downtime instead of shutdown the old connection, has a period of downtime until new connection is established.

@bart0sh back to the very initial question

> What would happen if Kubelet restarts in the middle of this process

I feel this is an existing issue, and this PR doesn't make thing worse and has a specific issue to resolve (supporting HA). Can we narrow down the scope of this PR and fix one thing at a time to have progressive improvement as long as there is no tech debt?

Comment (pohly) [reply pkg/kubelet/cm/devicemanager/manager_test.go:325]: I would prefer to have one PR which addresses this problem once and for all. Otherwise we have to review and maintain tests for a solution that might get replaced again.

We still have time before the 1.33 code freeze...

Comment (pohly) [reply pkg/kubelet/cm/devicemanager/manager_test.go:325]: The other consideration also is what guidance we provide to plugin authors. I'd rather have one "take this action now" for 1.33 than a "you can do this for 1.33, but expect to do something else soon".

Let's take a step back. If we really have the situation that two daemonset pods register under the same plugin name, then that situation is going to resolve itself pretty quickly. In case of a daemonset delete + re-create, the old pod lost its owner and automatically gets deleted. In case of a rolling update, the daemonset controller will eventually ensure that only one pod runs.

What if we let the kubelet track all potential endpoints for a certain plugin name and have it pick the one registered most recently when it needs to interact with that plugin? For a rolling update, it will use the newer one immediately. In the edge case of a kubelet restart, it might pick the older one for a short while until that one goes away. This seems acceptable to me.

The difference compared to the current solution in this PR is that the kubelet must track multiple endpoints. Otherwise we get the problem from the start of this thread (kubelet registers old plugin last, that goes away, no plugin left).

For this to work, plugins must use unique Unix domain socket paths for each pod. They cannot clean up for other pods because the other pod might still be running and the new one might fail to start properly. Each pod should try hard to clean up after itself.

Neither can all pods use the same path because then ownership of that one is ambiguous. It believe it is technically possible to have two separate processes listening on the same Unix domain socket, but that won't work well when the older one removes it from the filesystem on shutdown.

Plugins also must be written so that they can run in parallel. Can the kubelet inform the "active" instance when it has been selected and when it lost that status again? This might be a difference compared to the current registration process.

In the case of a kubelet restart and the newer plugin being registered first, it goes from "registered + active", to "on stand-by" when the older plugin gets registered, back to "active" when the older plugin goes away.

There is no "active" in https://github.com/kubernetes/kubernetes/blob/a78983906f4cf6f94029a3e6cf3171ef42b395d7/staging/src/k8s.io/kubelet/pkg/apis/pluginregistration/v1/api.proto#L41-L47

We could use that status such that a plugin gets `RegistrationStatus{PluginRegistered: true}` when it gets activated and `RegistrationStatus{PluginRegistered: false}` when some other instance takes over. This is different from how that API is currently used, so it has to be considered an API break. Plugins could opt into it via a new field in `PluginInfo`. That's fully backward compatible: older kubelets ignore it, so this new behavior simply won't work.

This calls for a KEP...

Comment (pohly) [reply pkg/kubelet/cm/devicemanager/manager_test.go:325]: To determine whether "registered" -> "unregistered" -> "registered" is an API break, let's look at how current plugins would react to that.

For CSI, registration is typically handled by the [node-driver-registrar](https://github.com/kubernetes-csi/node-driver-registrar/), running as a [CSI sidecar container](https://github.com/kubernetes-csi/csi-driver-host-path/blob/0bfbc4b4a466aabf945c51c40350c904a96a3459/deploy/kubernetes-1.27/hostpath/csi-hostpath-plugin.yaml#L278-L301). If that sidecar is told that it is "not registered", it [will exit](https://github.com/kubernetes-csi/node-driver-registrar/blob/27700e2962cd35b9f2336a156146181e5c75399e/cmd/csi-node-driver-registrar/main.go#L109-L112). That it turn causes a restart, so the "unregistered" plugin will register again, using [*the same* socket path](https://github.com/kubernetes-csi/node-driver-registrar/blob/27700e2962cd35b9f2336a156146181e5c75399e/cmd/csi-node-driver-registrar/node_register.go#L89-L91). So in the case of a rolling update, we end up with two pods fighting over who remains registered until one pod stops running.

For the proposed change that means it won't make things worse. We never get to sending "registered (again)" because the plugin immediately exits. The result of `NotifyRegistrationResult(registered=false)` is a timeout because no response ever gets sent.

For DRA, all (?!) current DRA drivers use https://github.com/kubernetes/dynamic-resource-allocation/blob/d47c7536d72ada0b2ce782eab2f1e0e582d96389/kubeletplugin/registrationserver.go#L46-L54. It returns an error to the kubelet when it receives a "not registered", but keeps running. This currently doesn't matter because [when the kubelet calls `og.notifyPlugin(... false)`](https://github.com/kubernetes/kubernetes/blob/f6f06806cc43ed9f7eb2f68368c90a8239884118/pkg/kubelet/pluginmanager/operationexecutor/operation_generator.go#L97-L127), there was already something wrong. In the future, we probably want that result to indicate "successfully deactivated". When we add code for replacing one instance with another, we should wait for success before sending the new instance a "you are registered". Otherwise we cannot guarantee that there is only one active instance, which IMHO is an important property.

TLDR: I'm arriving at the conclusion that "registered" -> "unregistered" -> "registered"  is backwards-compatible. To support it, plugins must be updated to handle "unregistered" differently. Does this sound right?

BTW, https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/pluginmanager/pluginwatcher/README.md seems to be the only documentation for this entire process. Whatever PR makes changes to this process (including this PR) should update that file.

File: pkg/kubelet/cm/devicemanager/plugin/v1beta1/handler.go (hunk 2 above)
Comment (pohly) [reply pkg/kubelet/cm/devicemanager/plugin/v1beta1/handler.go:116]: "ID returns the identifier of the client." doesn't actually explain what that ID *is*. Where does it come from? What properties does it have (for example, uniqueness)?

File: pkg/kubelet/cm/devicemanager/manager_test.go (hunk 4 above)
Comment (pohly) [reply pkg/kubelet/cm/devicemanager/manager_test.go:325]: > So in the case of a rolling update, we end up with two pods fighting over who remains registered until one pod stops running.

Okay, "devil's advocate" time. How much of a problem is that in practice? In other words, do we really need to do anything beyond what is proposed in this PR (keep the plugin registered as the new one registers)?

File: pkg/kubelet/cm/devicemanager/manager.go [hunk 5]
```diff
@@ -235,7 +235,12 @@ func (m *ManagerImpl) PluginConnected(resourceName string, p plugin.DevicePlugin
 
 	m.mutex.Lock()
 	defer m.mutex.Unlock()
-	m.endpoints[resourceName] = endpointInfo{e, options}
+	newEndpoint := endpointInfo{e, options}
+	if ep, exists := m.endpoints[resourceName]; exists {
+		klog.V(2).InfoS("Endpoint is taken over", "resourceName", resourceName, "prevEndpoint", ep, "newEndpoint", newEndpoint)
+		ep.e.setStopTime(time.Now())
+	}
+	m.endpoints[resourceName] = newEndpoint
```
Comment (pohly) [inline pkg/kubelet/cm/devicemanager/manager.go:243]: One thing that I don't see called out in this PR at the moment is whether plugins should or should not use the same socket path for each instance.

@linxiulei: can you clarify?

File: pkg/kubelet/cm/devicemanager/manager.go [hunk 6]
```diff
@@ -235,7 +235,12 @@ func (m *ManagerImpl) PluginConnected(resourceName string, p plugin.DevicePlugin
 
 	m.mutex.Lock()
 	defer m.mutex.Unlock()
-	m.endpoints[resourceName] = endpointInfo{e, options}
+	newEndpoint := endpointInfo{e, options}
```
Comment (pohly) [inline pkg/kubelet/cm/devicemanager/manager.go:238]: Not really related to this line, but I wanted to start a new discussion thread...

@ffromani agreed with me:
> FWIW I'm not a fan of the dual-socket model either. The work @SergeyKanzhelev started about device plugins reliability touches also this aspect. Unfortunately I'm skeptical we can change the model without breaking compability, but I'd be interested in alternatives, and willing to help as much as I can (likewise for the device plugin reliability effort)

What if a plugin creates only a single socket, serves the registration interface *and* the plugin interface (device/CSI/DRA) there, and simply announces the path of that socket to the kubelet when registering?

The kubelet doesn't need to know or care about this. When it sees the new socket, it starts registration. Then it connects again to the same socket. Nothing wrong with that, as far as I can see.

It doesn't simplify the kubelet code, but plugins become simpler.

File: pkg/kubelet/cm/devicemanager/manager_test.go (hunk 4 above)
Comment (bart0sh) [reply pkg/kubelet/cm/devicemanager/manager_test.go:325]: @linxiulei 
> case 2
> new plugin is connected first
> old plugin then takes over
> old plugin is tear down, and Kubelet would release the resource
> Kubelet effectively ignore the new plugin and the resource capacity is zero

This can be changed. I'm currently working on the PR https://github.com/kubernetes/kubernetes/pull/129799 that monitors registered sockets and unregisters plugins if they don't respond to `GetInfo` requests. However, it's possible to also make it to register unregistered plugins if they respond to `GetInfo` requests. It means that the new plugin from your case will be discovered quite soon after old plugin is teared down and registered.

File: pkg/kubelet/cm/devicemanager/manager.go (hunk 6 above)
Comment (pohly) [reply pkg/kubelet/cm/devicemanager/manager.go:238]: There might be one (very small) justification for the dual-socket model: plugins can be set up so that have shared access to the `plugin_registry` directory and only have access to their own private socket directory.

This prevents one plugin from impersonating the kubelet by connecting to some other plugin's socket and e.g. calling `NodeUnprepareResources`.

They may or may not be able to connect to other registration sockets. If they can, they can call `NotifyRegistrationResult(registered=false)` which is still bad.

What they probably can (as they need write permission for `plugin_registry`) is unlink some other driver's socket and thus cause it to be unregistered.

So even with the dual-socket model there are plenty of possibilities for one rogue plugin to cause havoc. Unless someone can demonstrate how to lock all of that down, I still don't see what the dual-socket model really provides.

File: pkg/kubelet/cm/devicemanager/manager_test.go (hunk 4 above)
Comment (bart0sh) [reply pkg/kubelet/cm/devicemanager/manager_test.go:325]: @pohly 
> If each pod deletes all other sockets except its own, we end up with kubelet talking to the old pod and once that finally goes away, there's no active plugin anymore.
> Can that happen? If not, why?

I'm not 100% sure, but I think a plugin would fail (exit grpc.Serve) if socket file is removed and in most cases it would also exit and will be restarted and will try to create its socket again.

File: pkg/kubelet/cm/devicemanager/manager.go (hunk 5 above)
Comment (bart0sh) [reply pkg/kubelet/cm/devicemanager/manager.go:243]: I think they shouldn't because they can't listen on the same unix socket. This is possible for TCP sockets in `SO_REUSEADDR` mode, but `SO_REUSEADDR` doesn't have any effect on unix sockets as far as I know.

File: pkg/kubelet/cm/devicemanager/manager.go (hunk 6 above)
Comment (bart0sh) [reply pkg/kubelet/cm/devicemanager/manager.go:238]: Would using one socket require adding [registration APIs](https://github.com/kubernetes/kubernetes/tree/master/staging/src/k8s.io/kubelet/pkg/apis/pluginregistration) to CSI, DRA and Device plugin APIs?

File: pkg/kubelet/cm/devicemanager/manager_test.go (hunk 4 above)
Comment (pohly) [reply pkg/kubelet/cm/devicemanager/manager_test.go:325]: > I'm not 100% sure, but I think a plugin would fail (exit grpc.Serve) if socket file is removed

No. The listening socket continues to exist, it is just not accessible through the file system anymore. The plugin won't notice that the socket directory entry got removed unless it actively watches via inotify or checks periodically - neither DRA nor CSI drivers do that.

File: pkg/kubelet/cm/devicemanager/manager.go (hunk 6 above)
Comment (pohly) [reply pkg/kubelet/cm/devicemanager/manager.go:238]: No.

I tried out this idea, it's entirely local to how the plugin manages it's sockets or rather, one socket. No API changes necessary.

Comment (pohly) [reply pkg/kubelet/cm/devicemanager/manager.go:238]: See https://github.com/kubernetes/kubernetes/pull/129809.

I'm a bit undecided because of the attack surface thing. It definitely is easier, but also a bit less secure.

File: pkg/kubelet/cm/devicemanager/manager.go (hunk 5 above)
Comment (linxiulei) [reply pkg/kubelet/cm/devicemanager/manager.go:243]: This is up to plugin authors/owners. If they wish to have HA for their plugin, they should use unique socket paths in  because HA requires starting new Pod before stopping the old Pod. In other words, with the rolling-update, if `maxSurge` is 0, the socket path could be identical as there is only one Pod running on a node during an update.

File: pkg/kubelet/cm/devicemanager/manager_test.go (hunk 4 above)
Comment (linxiulei) [reply pkg/kubelet/cm/devicemanager/manager_test.go:325]: > I would prefer to have one PR which addresses this problem once and for all. Otherwise we have to review and maintain tests for a solution that might get replaced again.

Agreed in principle on two considerations:

1. Does this PR bring in compatibility to future changes
2. When will the replacement be reasonably landed

> We still have time before the 1.33 code freeze...

I am more than happy to see an PR/KEP/what-is-needed in process that will be merged before 1.33 code freeze. Problem solved in one way or another : )

Comment (linxiulei) [reply pkg/kubelet/cm/devicemanager/manager_test.go:325]: @bart0sh https://github.com/kubernetes/kubernetes/pull/129799 seems to only monitor registered plugins right now. It should be helpful to also monitor the unregistered plugins whereas their sockets exist, as you suggested, to achieve eventual consistency in case of some race conditions.

File: pkg/kubelet/cm/devicemanager/manager.go (hunk 5 above)
Comment (pohly) [reply pkg/kubelet/cm/devicemanager/manager.go:243]: Agreed. This is something that we need to tell developers about. Making this code change and then not having any device plugin make use of it would be pointless.

File: pkg/kubelet/cm/devicemanager/manager_test.go (hunk 4 above)
Comment (pohly) [reply pkg/kubelet/cm/devicemanager/manager_test.go:325]: @linxiulei: that's a good point. Can you perhaps review that PR and comment there?

I think we'll have to figure something out among us three. I doubt that others will have the time or motivation anytime soon.

Comment (bart0sh) [reply pkg/kubelet/cm/devicemanager/manager_test.go:325]: @linxiulei Thanks for reminder. I'm thinking about it. It would probably be safer to only consider previously registered plugins so new code wouldn't race with existing `pluginwatcher` [reaction](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/pluginmanager/pluginwatcher/plugin_watcher.go#L75-L79) to socket creation.

**Comment by pohly** [comment]
I've tried to capture my understanding of the current state of affairs and how to move forward with the plugin watcher in https://github.com/kubernetes/kubernetes/pull/129832.

All PRs making changes in that space should edit `pkg/kubelet/pluginmanager/pluginwatcher/README.md`, so we will have to be careful in which order we merge PRs to minimize conflicts.

File: pkg/kubelet/cm/devicemanager/manager_test.go (hunk 4 above)
Comment (bart0sh) [reply pkg/kubelet/cm/devicemanager/manager_test.go:325]: @linxiulei @pohly 
> It should be helpful to also monitor the unregistered plugins whereas their sockets exist, as you suggested

Implemented. PTAL: https://github.com/kubernetes/kubernetes/pull/129799
